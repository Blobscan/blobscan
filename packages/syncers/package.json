{
  "name": "@blobscan/syncers",
  "description": "Blobscan's stats syncer",
  "private": true,
  "version": "0.3.2",
  "main": "./src/index.ts",
  "scripts": {
    "clean": "rm -rf .turbo node_modules",
    "lint": "eslint .",
    "lint:fix": "pnpm lint --fix",
    "type-check": "tsc --noEmit",
    "test": "pnpm with-env:test vitest",
    "test:ui": "pnpm with-env:test vitest --ui",
    "with-env:test": "dotenv -e ../../.env.test --"
  },
  "dependencies": {
    "@blobscan/dates": "workspace:*",
    "@blobscan/dayjs": "workspace:^0.1.0",
    "@blobscan/db": "workspace:^0.12.0",
    "@blobscan/logger": "workspace:^0.1.1",
    "@blobscan/zod": "workspace:^0.1.0",
    "axios": "^1.7.2",
    "bullmq": "^4.13.2",
    "ioredis": "^5.3.2"
  },
  "eslintConfig": {
    "root": true,
    "extends": [
      "@blobscan/eslint-config/base"
    ]
  }
}
