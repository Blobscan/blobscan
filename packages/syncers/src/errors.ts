export class ErrorException extends Error {
  constructor(message: string, cause?: unknown) {
    super(message, {
      cause,
    });

    this.name = this.constructor.name;
  }
}

export class SyncerError extends ErrorException {
  constructor(syncerName: string, message: string, cause: unknown) {
    super(`Syncer "${syncerName}" failed: ${message}`, cause);
  }
}
