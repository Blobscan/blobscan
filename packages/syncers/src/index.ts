export { BaseSyncer } from "./BaseSyncer";
export { createRedisConnection } from "./utils";

export { syncSwarmStamp } from "./syncers/SwarmStampSyncer";
export { aggregateDailyStats } from "./syncers/DailyStatsSyncer";
export { aggregateOverallStats } from "./syncers/OverallStatsSyncer";
