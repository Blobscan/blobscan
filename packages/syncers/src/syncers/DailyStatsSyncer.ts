import { normalizeDate, toDailyDate } from "@blobscan/dayjs";
import { prisma } from "@blobscan/db";
import { createModuleLogger } from "@blobscan/logger";

import { formatDate } from "../utils";

const logger = createModuleLogger("daily-stats-syncer");

export async function aggregateDailyStats() {
  const lastIndexedBlock = await prisma.block.findLatest();

  if (!lastIndexedBlock) {
    logger.debug("Skipping stats aggregation. No blocks indexed yet");
    return;
  }

  const targetDate = normalizeDate(lastIndexedBlock.timestamp).subtract(
    1,
    "day"
  );

  const targetDay = toDailyDate(targetDate);

  const rawLastDailyStatsDay = await prisma.dailyStats.findFirst({
    select: { day: true },
    where: { category: null, rollup: null },
    orderBy: { day: "desc" },
  });

  const lastDailyStatsDay = rawLastDailyStatsDay?.day
    ? normalizeDate(rawLastDailyStatsDay.day)
    : undefined;

  if (lastDailyStatsDay ? lastDailyStatsDay?.isSame(targetDay, "day") : false) {
    logger.debug(`Skipping stats aggregation. Already up to date`);
    return;
  }

  const res = await prisma.dailyStats.aggregate({
    from: lastDailyStatsDay?.add(1, "day"),
    to: targetDay,
  });

  logger.info(
    `Daily data up to day ${formatDate(
      targetDay
    )} aggregated. ${res} daily stats created successfully.`
  );
}
