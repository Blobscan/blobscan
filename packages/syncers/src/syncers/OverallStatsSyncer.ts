import { prisma } from "@blobscan/db";
import type { BlockNumberRange } from "@blobscan/db";
import { createModuleLogger } from "@blobscan/logger";

type OverallStatsSyncerConfig = {
  batchSize?: number;
  lowestSlot?: number;
};

const DEFAULT_BATCH_SIZE = 2_000_000;
const DEFAULT_INITIAL_SLOT = 0;

function isUnset<T>(value: T | undefined | null): value is null | undefined {
  return value === undefined || value === null;
}

const logger = createModuleLogger("overall-stats-syncer");

export async function aggregateOverallStats({
  batchSize = DEFAULT_BATCH_SIZE,
  lowestSlot = DEFAULT_INITIAL_SLOT,
}: OverallStatsSyncerConfig) {
  const [blockchainSyncState, latestBlock] = await Promise.all([
    prisma.blockchainSyncState.findUnique({
      select: {
        lastLowerSyncedSlot: true,
        lastAggregatedBlock: true,
        lastFinalizedBlock: true,
      },
      where: {
        id: 1,
      },
    }),
    prisma.block.findLatest(),
  ]);
  const { lastAggregatedBlock, lastFinalizedBlock, lastLowerSyncedSlot } =
    blockchainSyncState ?? {};
  const lastIndexedBlockNumber = latestBlock?.number;

  if (
    isUnset(lastIndexedBlockNumber) ||
    isUnset(lastFinalizedBlock) ||
    lastLowerSyncedSlot !== lowestSlot
  ) {
    logger.debug(
      "Skipping stats aggregation. Chain hasn't been fully indexed yet"
    );

    return;
  }

  const blockRange: BlockNumberRange = {
    from: lastAggregatedBlock ? lastAggregatedBlock + 1 : 0,
    to: Math.min(lastFinalizedBlock, lastIndexedBlockNumber),
  };

  if (blockRange.from >= blockRange.to) {
    logger.debug(
      `Skipping stats aggregation. No new finalized blocks (last aggregated block: ${lastAggregatedBlock})`
    );

    return;
  }

  const { from, to } = blockRange;
  const unprocessedBlocks = to - from + 1;
  const batches = Math.ceil(unprocessedBlocks / batchSize);

  for (let i = 0; i < batches; i++) {
    const batchFrom = from + i * batchSize;
    const batchTo = Math.min(batchFrom + batchSize - 1, to);
    const batchBlockRange: BlockNumberRange = {
      from: batchFrom,
      to: batchTo,
    };

    await prisma.overallStats.aggregate({ blockRange: batchBlockRange });

    if (batches > 1) {
      const formattedFromBlock = batchBlockRange.from.toLocaleString();
      const formattedToBlock = batchBlockRange.to.toLocaleString();

      logger.debug(
        `(batch ${
          i + 1
        }/${batches}) Data from block ${formattedFromBlock} up to ${formattedToBlock} aggregated successfully`
      );
    }
  }

  logger.info(
    `Data from block ${blockRange.from.toLocaleString()} up to ${blockRange.to.toLocaleString()} aggregated successfully.`
  );
}
