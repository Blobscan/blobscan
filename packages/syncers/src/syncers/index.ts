export { DailyStatsSyncer } from "./DailyStatsSyncer";
export type { DailyStatsSyncerConfig } from "./DailyStatsSyncer";
export { OverallStatsSyncer } from "./OverallStatsSyncer";
export type { OverallStatsSyncerConfig } from "./OverallStatsSyncer";
