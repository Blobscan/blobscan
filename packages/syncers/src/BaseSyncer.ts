import { Queue, Worker } from "bullmq";
import type { Redis } from "ioredis";

import { createModuleLogger } from "@blobscan/logger";
import type { Logger } from "@blobscan/logger";

import { ErrorException, SyncerError } from "./errors";

export type BaseSyncerConfig = {
  name: string;
  connection: Redis;
  cronPattern: string;
  syncerFn: () => Promise<void>;
};

export class BaseSyncer {
  private name: string;
  private cronPattern: string;
  private logger: Logger;

  public readonly worker: Worker;
  public readonly queue: Queue;

  constructor({ name, cronPattern, connection, syncerFn }: BaseSyncerConfig) {
    this.name = `${name}-syncer`;
    this.cronPattern = cronPattern;
    this.logger = createModuleLogger(this.name);

    this.queue = new Queue(this.name, { connection });
    this.worker = new Worker(this.queue.name, syncerFn, { connection });

    this.queue.on("error", (error) => {
      this.logger.error(new ErrorException("A queue error occurred", error));
    });

    this.worker.on("failed", (_, error) => {
      this.logger.error(new ErrorException("A worker error occurred", error));
    });
  }

  async start() {
    try {
      const repeatableJob = await this.queue.add(`${this.name}-job`, null, {
        repeat: {
          pattern: this.cronPattern,
        },
      });

      this.logger.info("Syncer started successfully");

      return repeatableJob;
    } catch (error) {
      throw new SyncerError(
        this.name,
        "An error occurred when starting syncer",
        error
      );
    }
  }

  async close(): Promise<void> {
    const results = await Promise.allSettled([
      this.queue.obliterate({ force: true }),
      this.queue.removeAllListeners().close(),
      this.worker.removeAllListeners().close(true),
    ]);

    for (const result of results) {
      if (result.status === "rejected") {
        throw new SyncerError(
          this.name,
          "An error occurred when performing closing operation",
          result.reason
        );
      }
    }

    this.logger.info("Syncer closed successfully");
  }
}
