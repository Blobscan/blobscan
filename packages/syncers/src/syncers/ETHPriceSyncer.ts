import type { Address } from "viem";
import { createPublicClient, http } from "viem";
import { mainnet } from "viem/chains";

import { prisma } from "@blobscan/db";
import { PriceFeedFinder } from "@blobscan/eth-price";

import { BaseSyncer } from "../BaseSyncer";
import type { CommonSyncerConfig } from "../BaseSyncer";

const ONE_HOUR_SECONDS = BigInt(60 * 60);

export interface ETHPriceSyncerConfig extends CommonSyncerConfig {
  chainJsonRpcUrl: string;
  ethUsdDataFeedContractAddress: string;
}

let ethUsdPriceFinder: PriceFeedFinder | undefined;

function getEthUsdPriceFinder(
  chainJsonRpcUrl: string,
  ethUsdDataFeedContractAddress: string
) {
  if (!ethUsdPriceFinder) {
    ethUsdPriceFinder = new PriceFeedFinder(
      createPublicClient({
        chain: {
          ...mainnet,
          rpcUrls: {
            default: {
              http: [chainJsonRpcUrl],
            },
          },
        },
        transport: http(),
      }),
      ethUsdDataFeedContractAddress as Address,
      ONE_HOUR_SECONDS
    );
  }

  return ethUsdPriceFinder;
}

export class ETHPriceSyncer extends BaseSyncer {
  constructor(config: ETHPriceSyncerConfig) {
    super({
      ...config,
      name: "eth-price",
      syncerFn: async () => {
        const ethUsdPriceFinder = getEthUsdPriceFinder(
          config.chainJsonRpcUrl,
          config.ethUsdDataFeedContractAddress
        );
        const priceData = await ethUsdPriceFinder.getPriceByTimestamp(
          BigInt(Math.floor(Date.now() / 1_000))
        );

        // The timestamp is a bigint in seconds.
        // Prisma expects a Date object.
        const timestamp = new Date(Number(priceData.timestampSeconds) * 1_000);

        await prisma.ethUsdPrice.create({
          data: {
            price: priceData.price.toString(),
            timestamp,
          },
        });

        this.logger.info(`Fetched price data: ${JSON.stringify(priceData)}`);
      },
    });
  }
}
