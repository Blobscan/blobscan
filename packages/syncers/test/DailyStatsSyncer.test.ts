import { beforeEach, describe, expect, it, vi } from "vitest";

import { getDateFromISODateTime, toDailyDate } from "@blobscan/dayjs";
import { prisma } from "@blobscan/db";

import { DailyStatsSyncer } from "../src/syncers/";
import { CURRENT_DAY_DATA } from "./DailyStatsSyncer.test.fixtures";
import {
  getAllDailyStatsDates,
  indexNewBlock,
} from "./DailyStatsSyncer.test.utils";

class DailyStatsSyncerMock extends DailyStatsSyncer {
  constructor(redisUri = process.env.REDIS_URI ?? "") {
    super({ redisUriOrConnection: redisUri, cronPattern: "* * * * *" });
  }

  getWorker() {
    return this.worker;
  }

  getWorkerProcessor() {
    return this.syncerFn;
  }

  getQueue() {
    return this.queue;
  }
}

describe("DailyStatsSyncer", () => {
  let dailyStatsSyncer: DailyStatsSyncerMock;

  beforeEach(() => {
    dailyStatsSyncer = new DailyStatsSyncerMock();

    return async () => {
      await dailyStatsSyncer.close();
    };
  });

  it("should aggregate data for all available days", async () => {
    const workerProcessor = dailyStatsSyncer.getWorkerProcessor();

    await workerProcessor();

    const { blobDailyStats, blockDailyStats, txDailyStats } =
      await getAllDailyStatsDates();

    expect(blobDailyStats, "Blob daily stats mismatch").toMatchSnapshot();
    expect(blockDailyStats, "Block daily stats mismatch").toMatchSnapshot();
    expect(txDailyStats, "Transaction daily stats mismatch").toMatchSnapshot();
  });

  it("should skip aggregation if not all blocks have been indexed for the last day", async () => {
    const workerProcessor = dailyStatsSyncer.getWorkerProcessor();

    await indexNewBlock(CURRENT_DAY_DATA);

    await workerProcessor();

    const { blobDailyStats, blockDailyStats, txDailyStats } =
      await getAllDailyStatsDates();

    const currentDay = getDateFromISODateTime(
      toDailyDate(CURRENT_DAY_DATA.block.timestamp)
    );

    expect(
      blobDailyStats.find(([day]) => day === currentDay),
      "Blob daily stats should not include the current day"
    ).toBeUndefined();
    expect(
      blockDailyStats.find((day) => day === currentDay),
      "Block daily stats should not include the current day"
    ).toBeUndefined();
    expect(
      txDailyStats.find(([day]) => day === currentDay),
      "Transaction daily stats should not include the current day"
    ).toBeUndefined();
  });

  it("should skip aggregation if no blocks have been indexed yet", async () => {
    const workerProcessor = dailyStatsSyncer.getWorkerProcessor();

    const findLatestSpy = vi
      .spyOn(prisma.block, "findLatest")
      .mockImplementationOnce(() => Promise.resolve(null));

    await workerProcessor();

    const { blobDailyStats, blockDailyStats, txDailyStats } =
      await getAllDailyStatsDates();

    expect(findLatestSpy, "findLatest should be called").toHaveBeenCalled();
    expect(blobDailyStats, "Blob daily stats should be empty").toEqual([]);
    expect(blockDailyStats, "Block daily stats should be empty").toEqual([]);
    expect(txDailyStats, "Transaction daily stats should be empty").toEqual([]);
  });

  it("should skip aggregation if already up to date", async () => {
    await indexNewBlock(CURRENT_DAY_DATA);

    const workerProcessor = dailyStatsSyncer.getWorkerProcessor();

    await workerProcessor();

    const blobPopulateSpy = vi.spyOn(prisma.blobDailyStats, "populate");

    const blockPopulateSpy = vi.spyOn(prisma.blockDailyStats, "populate");

    const transactionPopulateSpy = vi.spyOn(
      prisma.transactionDailyStats,
      "populate"
    );

    await workerProcessor();

    expect(
      blobPopulateSpy,
      "Blob daily stats should not be populated"
    ).not.toHaveBeenCalled();
    expect(
      blockPopulateSpy,
      "Block daily stats should not be populated"
    ).not.toHaveBeenCalled();
    expect(
      transactionPopulateSpy,
      "Transaction daily stats should not be populated"
    ).not.toHaveBeenCalled();

    blobPopulateSpy.mockRestore();
    blockPopulateSpy.mockRestore();
    transactionPopulateSpy.mockRestore();
  });
});
