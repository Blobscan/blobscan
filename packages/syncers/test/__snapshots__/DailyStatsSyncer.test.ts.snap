// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DailyStatsSyncer > should aggregate data for all available days 1`] = `
[
  [
    "2022-10-16",
    "OTHER",
    null,
  ],
  [
    "2022-10-16",
    "ROLLUP",
    null,
  ],
  [
    "2022-10-16",
    null,
    "OPTIMISM",
  ],
  [
    "2022-10-16",
    null,
    null,
  ],
  [
    "2023-05-10",
    "ROLLUP",
    null,
  ],
  [
    "2023-05-10",
    null,
    "BASE",
  ],
  [
    "2023-05-10",
    null,
    null,
  ],
  [
    "2023-08-03",
    "OTHER",
    null,
  ],
  [
    "2023-08-03",
    null,
    null,
  ],
  [
    "2023-08-20",
    "OTHER",
    null,
  ],
  [
    "2023-08-20",
    "ROLLUP",
    null,
  ],
  [
    "2023-08-20",
    null,
    "ARBITRUM",
  ],
  [
    "2023-08-20",
    null,
    "BASE",
  ],
  [
    "2023-08-20",
    null,
    "OPTIMISM",
  ],
  [
    "2023-08-20",
    null,
    null,
  ],
  [
    "2023-08-28",
    "OTHER",
    null,
  ],
  [
    "2023-08-28",
    "ROLLUP",
    null,
  ],
  [
    "2023-08-28",
    null,
    "OPTIMISM",
  ],
  [
    "2023-08-28",
    null,
    null,
  ],
]
`;

exports[`DailyStatsSyncer > should aggregate data for all available days 2`] = `
[
  "2022-10-16",
  "2023-05-10",
  "2023-08-03",
  "2023-08-20",
  "2023-08-28",
]
`;

exports[`DailyStatsSyncer > should aggregate data for all available days 3`] = `
[
  [
    "2022-10-16",
    "OTHER",
    null,
  ],
  [
    "2022-10-16",
    "ROLLUP",
    null,
  ],
  [
    "2022-10-16",
    null,
    "OPTIMISM",
  ],
  [
    "2022-10-16",
    null,
    null,
  ],
  [
    "2023-05-10",
    "ROLLUP",
    null,
  ],
  [
    "2023-05-10",
    null,
    "BASE",
  ],
  [
    "2023-05-10",
    null,
    null,
  ],
  [
    "2023-08-03",
    "OTHER",
    null,
  ],
  [
    "2023-08-03",
    null,
    null,
  ],
  [
    "2023-08-20",
    "OTHER",
    null,
  ],
  [
    "2023-08-20",
    "ROLLUP",
    null,
  ],
  [
    "2023-08-20",
    null,
    "ARBITRUM",
  ],
  [
    "2023-08-20",
    null,
    "BASE",
  ],
  [
    "2023-08-20",
    null,
    "OPTIMISM",
  ],
  [
    "2023-08-20",
    null,
    null,
  ],
  [
    "2023-08-28",
    "OTHER",
    null,
  ],
  [
    "2023-08-28",
    "ROLLUP",
    null,
  ],
  [
    "2023-08-28",
    null,
    "OPTIMISM",
  ],
  [
    "2023-08-28",
    null,
    null,
  ],
]
`;
